src/nnbench/types.py

```diff
@@ -93,9 +94,24 @@ class Benchmark:
     tearDown: Callable[..., None] = field(repr=False, default=NoOp)
     tags: tuple[str, ...] = field(repr=False, default=())
 
+    @dataclass(frozen=True)
+    class Interface:
+        fn: Callable[..., Any]
+        varnames: Tuple[str, ...] | None = field(default=None)
+        vartypes: Tuple[inspect.Parameter, ...] | None = field(default=None)
+        varitems: Tuple[Tuple[str, inspect.Parameter], ...] | None = field(default=None)
+        defaults: Dict[str, Any] | None = field(default=None)
+
+        def __post_init__(self) -> None:
+            sig = inspect.signature(self.fn)
+            super().__setattr__("varnames", tuple(sig.parameters.keys()))
+            super().__setattr__("vartypes", tuple(sig.parameters.values()))
+            super().__setattr__("varitems", tuple(sig.parameters.items()))
+            super().__setattr__("defaults", {n: p.default for n, p in sig.parameters.items()})
```

Comment:
This together with `fn` appearing in the initializer is a bit unfortunate. How about we copy `inspect.Signature` in that regard by using a classmethod?

```py
@dataclass(frozen=True)
class Interface:
    # no fn slot here.
    varnames: tuple[str, ...]  # <- these should never be None.
    vartypes: tuple[type, ...] # <- ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...

    @classmethod
    def from_callable(cls, fn: Callable) -> "Interface":
        ... # fill all of the above slots.
        return cls(...)
```

Notice the `tuple[type, ...]` on `vartypes` - the point is to grab the metadata from the `Parameter`s, not save the parameters themselves (otherwise we would just keep rolling with the `inspect.Signature`!)

I would even go so far as to not make it default-constructible (i.e. leave out `__init__`), since this is internal-only and we're just refining some of the info from the `inspect.Signature`.

Reply:
I just realized I screwed up the `vartypes` typing in the issue, so that's on me.